Transcoding from format with rate <> 16000 can cause choppy sound.

mtryfoss opened the pull request:
Timestamp on outgoing packets are increased by rate of incoming stream. For example steps of 160 instead of 320 when transcoding from sln8.

This is not triggered when transcoding SIP-SIP because of Asterisk nature, but can be triggered by localy generated audio in for example sln8 format.

mtryfoss committed:
Transcoding from format with rate <> 16000 can cause choppy sound. Timestamp on outgoing packets are increased by rate of incoming stream. For example steps of 160 instead of 320 when transcoding from sln8.
This is not triggered when transcoding SIP-SIP because of Asterisk nature, but can be triggered by localy generated audio in for example sln8 format.

mtryfoss commented:
Will come back with a new. This is wrong for other rates.

traud commented:
Thanks for reporting (and thanks for upvoting via a thumbs up). Yes, with the current code, the timestamp of outgoing RTP packets differs not always ‘320’ but depends on the source audio-codec (`src_codec`) of the currently active `ast_translator`. Finally, I was able to reproduce this, for example:

**Steps to Reproduce**

1. comment `lin16toevs` (and `evstolin16`) in the method `load_module` (and `unload_module`) of the source file `codecs/codec_evs.c`
2. with an 3GPP-EVS enabled client, call an extension which plays a static file, for example in the configuration file `/etc/asterisk/extensions.conf` go for `exten => 600,1,Playback(demo-echotest)`. 
3. open Wireshark and monitor the RTP packets leaving Digium Asterisk

**Expected Results**

In Wireshark, the RTP timestamp increases at 320. Note: Because the Asterisk demo files do not have much speech pauses, the usual [DTX](https://en.wikipedia.org/wiki/Discontinuous_transmission) does not kick in, which would confuse the timestamp.

**Actual Results**

In Wireshark, the RTP timestamp increases at 160 because the translator `evstolin` is active, which bridges the 8000 kHz variants of the Asterisk demo files (‘sln8’).

**Note 1**
[3GPP TS 26.445](http://www.3gpp.org/DynaReport/26445.htm) Annex A.1 states ‘The timestamp clock frequency for EVS codec is 16 kHz regardless of audio bandwidth.’ I am unsure why I introduced this software bug. None of the Asterisk included and not one of my other transcoding modules behave that way. Consequently, this was not a simple copy and paste error. Especially the module Opus Codec should have been the role model here because its RFC 7587 section 4.1 states ‘The RTP timestamp is incremented with a 48000 Hz clock rate for all modes of Opus and all sampling rates.’ And I [was aware](/traud/asterisk-opus/issues/10) of implementations violating that statement. Therefore, why this software bug happened to me is obscure.

**Note 2**
Not one of my EVS clients played choppy sound. Not sure why. Either they ignore the timestamp or they see other values. Here, Wireshark was required to see the issue.

> This is not triggered when transcoding SIP-SIP […]

Able to confirm that as well. In the source file `res/res_rtp_asterisk.c`, in the method `rtp_raw_write`, the code contains something like
```
	if (ast_test_flag(frame, AST_FRFLAG_HAS_TIMING_INFO)) {
		rtp->lastts = frame->ts * (rtp_get_rate(frame->subclass.format) / 1000);
	}
```
That overrides my (wrong) value supplied to `ast_trans_frameout`. That flag is set in `ast_rtp_read`. According to the source code comments because of jitter buffer handling. In other words, that flags is set on the other call leg and survives not only the jitter buffer of the other call leg but also the transcoding on the current call leg. Therefore this bug here was hidden by that. I do not think this is correct behavior either. Looks like, you unveiled another issue. :-)

traud committed:
fix RTP timestamp to avoid choppy audio

3GPP TS 26.445 Annex A.1: ‘The timestamp clock frequency for EVS codec is 16 kHz regardless of audio bandwidth.’
